### PR TITLE
Add support for OpenSSL 3

### PR DESCRIPTION
**Pull request by adfoster-r7**
Relates to https://github.com/rapid7/metasploit-framework/issues/16818

Updates metasploit-framework to work with OpenSSL 3.0. Let's also run Ubuntu 22.04 in our test matrix, to catch any OpenSSL v3 regression issues

## Verification

- Code review
- Verify CI passes

### CODE REVIEW

File: lib/msf/core/exploit/remote/kerberos/client/as_request.rb [hunk 1]
```diff
@@ -44,7 +44,7 @@ def build_as_pa_time_stamp(opts = {})
               time_stamp = opts[:time_stamp] || Time.now
               pausec = opts[:pausec] || 0
               etype = opts[:etype] || Rex::Proto::Kerberos::Crypto::RC4_HMAC
-              key = opts[:key] || ''
+              key = opts[:key] || OpenSSL::Random.random_bytes(16)
```
Comment (adfoster-r7) [inline lib/msf/core/exploit/remote/kerberos/client/as_request.rb:47]: OpenSSL 3 throws errors on invalid keys

File: Gemfile [hunk 2]
```diff
@@ -3,6 +3,12 @@ source 'https://rubygems.org'
 #   spec.add_runtime_dependency '<name>', [<version requirements>]
 gemspec name: 'metasploit-framework'
 
+gem 'ruby_smb', github: 'adfoster-r7/ruby_smb', branch: 'run-ubuntu-22.04-in-test-matrix'
+gem 'hrr_rb_ssh', github: 'adfoster-r7/hrr_rb_ssh', branch: 'investigate-openssl3-support'
+gem 'openssl-ccm', github: 'adfoster-r7/openssl-ccm', branch: 'add-support-openssl-3'
+gem 'openssl-cmac', github: 'adfoster-r7/openssl-cmac', branch: 'add-support-for-openssl3'
+gem 'metasploit-credential', github: 'adfoster-r7/metasploit-credential', branch: 'add-support-for-openssl3'
+
```
Comment (adfoster-r7) [inline Gemfile:11]: These custom forks will either need to be shipped by the maintainers, or we'll have to temporarily move them under the Rapid7 Github account until they're merged upstream and officially available in RubyGems

**Comment by adfoster-r7** [comment]
@msjenkins-r7 retest this please

File: lib/rex/proto/rfb/cipher.rb [hunk 3]
```diff
@@ -83,14 +83,30 @@ def self.encrypt_ard(username, password, generator, key_length, prime_modulus, p
...
 
-    dh_peer = OpenSSL::PKey::DH.new(key_length * 8, generator)
-    dh_peer.set_key(peer_public_key, nil)
-
-    dh = OpenSSL::PKey::DH.new(dh_peer)
-    dh.set_pqg(prime_modulus, nil, generator)
-    dh.generate_key!
-
-    shared_key = dh.compute_key(dh_peer.pub_key)
+    # OpenSSL 3.0+
```
Comment (adfoster-r7) [inline lib/rex/proto/rfb/cipher.rb:86]: Code path triggered via vnc_login and connecting to an apple VNC server after manually adding a user that has permissions to log in, as well as going into Mac's sharing menu, enabling `Screen Sharing` and granting access to the new `test_user` account

```
msf6 auxiliary(scanner/vnc/vnc_login) > run 192.168.123.1 username=test_user password=p4$$w0rd1

[*] 192.168.123.1:5900    - 192.168.123.1:5900 - Starting VNC login sweep
[!] 192.168.123.1:5900    - No active DB -- Credential data will not be saved!
[+] 192.168.123.1:5900    - 192.168.123.1:5900 - Login Successful: test_user:p4$$w0rd1
[*] 192.168.123.1:5900    - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

File: Gemfile.lock [hunk 4]
```diff
@@ -502,19 +532,24 @@ PLATFORMS
...
   rubocop
   ruby-prof (= 1.4.2)
+  ruby_smb!
   simplecov (= 0.18.2)
   timecop
   yard
 
 BUNDLED WITH
-   2.1.4
+   2.2.22
```
Comment (adfoster-r7) [inline Gemfile.lock:555]: This shouldn't be here when we get the upstream forks landed

**Comment by adfoster-r7** [comment]
@msjenkins-r7 retest this please

File: Gemfile [hunk 5]
```diff
@@ -3,6 +3,12 @@ source 'https://rubygems.org'
 #   spec.add_runtime_dependency '<name>', [<version requirements>]
 gemspec name: 'metasploit-framework'
 
+# gem 'ruby_smb', github: 'adfoster-r7/ruby_smb', branch: 'run-ubuntu-22.04-in-test-matrix'
+# gem 'hrr_rb_ssh', github: 'adfoster-r7/hrr_rb_ssh', branch: 'investigate-openssl3-support'
```
Comment (adfoster-r7) [inline Gemfile:7]: I'll split this out into a separate PR, as it's blocked on an upstream PR merge and subsequent version bump

File: Gemfile (hunk 2 above)
Comment (adfoster-r7) [reply Gemfile:11]: These are all landed now, apart from `hrr_rb_ssh`

File: Gemfile.lock [hunk 6]
```diff
@@ -234,7 +234,7 @@ GEM
       activemodel (~> 6.0)
       activesupport (~> 6.0)
       railties (~> 6.0)
-    metasploit-credential (5.0.7)
+    metasploit-credential (5.0.8)
```
Comment (adfoster-r7) [inline Gemfile.lock:237]: Adds support for OpenSSL3 https://github.com/rapid7/metasploit-credential/pull/163

File: Gemfile.lock [hunk 7]
```diff
@@ -287,8 +287,8 @@ GEM
     octokit (4.25.1)
       faraday (>= 1, < 3)
       sawyer (~> 0.9)
-    openssl-ccm (1.2.2)
-    openssl-cmac (2.0.1)
+    openssl-ccm (1.2.3)
+    openssl-cmac (2.0.2)
```
Comment (adfoster-r7) [inline Gemfile.lock:291]: Adds support for openssl 3 https://github.com/SmallLars/openssl-cmac/pull/5

File: Gemfile.lock [hunk 8]
```diff
@@ -287,8 +287,8 @@ GEM
     octokit (4.25.1)
       faraday (>= 1, < 3)
       sawyer (~> 0.9)
-    openssl-ccm (1.2.2)
-    openssl-cmac (2.0.1)
+    openssl-ccm (1.2.3)
```
Comment (adfoster-r7) [inline Gemfile.lock:290]: Adds support for openssl 3 https://github.com/SmallLars/openssl-ccm/pull/10

File: Gemfile.lock [hunk 9]
```diff
@@ -429,7 +429,7 @@ GEM
     ruby-progressbar (1.11.0)
     ruby-rc4 (0.1.5)
     ruby2_keywords (0.0.5)
-    ruby_smb (3.1.6)
+    ruby_smb (3.1.7)
```
Comment (adfoster-r7) [inline Gemfile.lock:432]: Adds support for openssl 3 https://github.com/rapid7/ruby_smb/pull/234

File: lib/msfenv.rb [hunk 10]
```diff
@@ -1,6 +1,11 @@
 # Use bundler to load dependencies
 #
 
+# Enable legacy providers such as blowfish-cbc, cast128-cbc, arcfour, etc
+ENV['OPENSSL_CONF'] = File.expand_path(
```
[the remark this answers is not in the document]
Comment (jmartin-r7) [reply lib/msfenv.rb:5]: Per the [latest docs the syntax](https://www.openssl.org/docs/man3.0/man5/config.html) in the file can be parsed differently by older versions.  Specifically there are notes about `=` requirements changing semantics.

Given there are ways this config file can manipulate the runtime paths I am not sure if overriding the user's env without offering control or a warning is the best action to take.

[the remark this answers is not in the document]
Comment (adfoster-r7) [reply lib/msfenv.rb:5]: I'm not strongly opinionated here, let me know what you'd prefer and I can implement it :+1:

Looks like net-ssh uses this pattern:

```ruby
ENV['OPENSSL_CONF'] = 'test/openssl3.conf' if OpenSSL::OPENSSL_LIBRARY_VERSION.start_with? "OpenSSL 3"
```

[the remark this answers is not in the document]
Comment (jmartin-r7) [reply lib/msfenv.rb:5]: Not 100% sure if it should just warn or suppress when we detect an existing config.  Are you good with starting with a warning and we can change later if we get user feedback that clobbering the env is problematic for them?

```Ruby
if OpenSSL::OPENSSL_LIBRARY_VERSION.start_with? "OpenSSL 3"
  puts "Overriding user environment variable 'OPENSSL_CONF' to enable legacy functions." unless ENV['OPENSSL_CONF'].blank?
  ENV['OPENSSL_CONF'] = File.expand_path(
...
end
```

Also not sure if `puts` is best option or if we have some other mechanism to warn and log.

[the remark this answers is not in the document]
Comment (adfoster-r7) [reply lib/msfenv.rb:5]: Ah, there's also the issue of setting the ENV after loading OpenSSL doesn't work, so it's not possible to check the OpenSSL version in a guard clause.

### Loading OpenSSL then setting env

Code:

```ruby
require 'openssl'
if OpenSSL::OPENSSL_LIBRARY_VERSION.start_with?('OpenSSL 3')
  $stderr.puts "Overriding user environment variable 'OPENSSL_CONF' to enable legacy functions." unless ENV['OPENSSL_CONF'].nil?
  ENV['OPENSSL_CONF'] = '/mnt/hgfs/metasploit-framework/config/openssl.conf'
end

OpenSSL::Cipher.new('BF-CBC')
```

Output:
```
$ ruby ./test.rb
./test.rb:7:in `initialize': unsupported (OpenSSL::Cipher::CipherError)
	from ./test.rb:7:in `new'
	from ./test.rb:7:in `<main>'
```

### Setting env then loading OpenSSL

Code:
```ruby
# require 'openssl'
# if OpenSSL::OPENSSL_LIBRARY_VERSION.start_with?('OpenSSL 3')
  $stderr.puts "Overriding user environment variable 'OPENSSL_CONF' to enable legacy functions." unless ENV['OPENSSL_CONF'].nil?
  ENV['OPENSSL_CONF'] = '/mnt/hgfs/metasploit-framework/config/openssl.conf'
# end

require 'openssl'
puts OpenSSL::Cipher.new('BF-CBC')
```

Output:
```
$ ruby ./test.rb
#<OpenSSL::Cipher:0x000055d82db29098>
```

---

So I believe the best I can do is continue using the current pattern, but with the extra log statement present:

```ruby
$stderr.puts "Overriding user environment variable 'OPENSSL_CONF' to enable legacy functions." unless ENV['OPENSSL_CONF'].nil?
ENV['OPENSSL_CONF'] = File.expand_path(
  File.join(File.dirname(__FILE__), '..', 'config', 'openssl.conf')
)
```

And swapping the suggestion of `blank?` for `nil?` - as ActiveSupport isn't available here

[the remark this answers is not in the document]
Comment (jmartin-r7) [reply lib/msfenv.rb:5]: Sounds like a reasonable compromise. Thank you for working through this.

[the remark this answers is not in the document]
Comment (adfoster-r7) [reply lib/msfenv.rb:6]: Updated in https://github.com/rapid7/metasploit-framework/pull/16800/commits/756c910b9ad7835f6e8c82da82202c8025eae2af :+1:

[the remark this answers is not in the document]
Comment (jmartin-r7) [reply lib/msfenv.rb:6]: Thank you, final testing in progress should land shortly.

**Comment by jmartin-r7** [comment]
### Release Notes
This adds support for OpenSSL 3 compatibility with legacy ciphers.

**Comment by adfoster-r7** [comment]
cc @sbrun for visibility :+1: 

This fixes a lot of the OpenSSL 3 issues with Metasploit on Kali, especially the SMB modules such as `windows/smb/ms17_010_eternalblue`/`auxiliary/admin/dcerpc/cve_2020_1472_zerologon` etc. The original Kali OpenSSL issue (https://bugs.kali.org/view.php?id=7804) was closed so I couldn't provide a progress update there.

**Pull request merged**